packages/doenetml-worker-rust/lib-doenetml-core/src/dep_graph/graph.rs

```diff
@@ -0,0 +1,307 @@
+
+    impl<'a, Node> DescendantTopologicalIterator<'a, Node> {
+        pub fn new(nodes: &'a [Node], edges: &'a [Vec<usize>], start_index: usize) -> Self {
+            let rti = ReverseTopologicalIterator::new(edges, start_index);
+            let mut order = rti.collect::<Vec<_>>();
+            order.reverse();
+
+            Self {
+                nodes,
+                iter: order.into_iter(),
+            }
+        }
+    }
+
+    impl<'a, Node> Iterator for DescendantTopologicalIterator<'a, Node> {
+        type Item = &'a Node;
+
+        fn next(&mut self) -> Option<Self::Item> {
+            self.iter.next().map(|index| &self.nodes[index])
+        }
+    }
+
+    /// An iterator that yields indices of all descendants of a node in a graph in reverse topological order.
+    pub struct ReverseTopologicalIterator<'a> {
+        edges: &'a [Vec<usize>],
+        remaining_indices: Vec<usize>,
+        visited: Vec<bool>,
+    }
+
+    impl<'a> ReverseTopologicalIterator<'a> {
+        pub fn new(edges: &'a [Vec<usize>], start_index: usize) -> Self {
+            let visited = vec![false; edges.len()];
+            let remaining_indices = edges[start_index].clone();
+            Self {
+                edges,
+                visited,
+                remaining_indices,
+            }
+        }
+    }
+
+    impl<'a> Iterator for ReverseTopologicalIterator<'a> {
+        type Item = usize;
+
+        fn next(&mut self) -> Option<Self::Item> {
+            // We want to return nodes sorted topologically.
+            while let Some(&index) = self.remaining_indices.last() {
+                // In the worst case, we have to queue the whole graph for iteration.
+                // If we have more queued than that, we have a cycle.
+                if self.remaining_indices.len() > self.edges.len() {
+                    panic!("Cycle detected in graph")
+                }
+
+                // If the node has been visited, pop it off the stack and continue.
+                if self.visited[index] {
+                    self.remaining_indices.pop();
+                    continue;
+                }
+                // If the node has no unvisited children, that's the node we want
+                // to return next.
+                if self.edges[index].iter().all(|&to| self.visited[to]) {
+                    self.visited[index] = true;
+                    self.remaining_indices.pop();
+                    return Some(index);
+                }
+                // Otherwise, add all unvisited children to the stack.
+                self.remaining_indices
+                    .extend(self.edges[index].iter().filter(|&&to| !self.visited[to]));
+            }
+            None
+        }
+    }
+}
+
+/// `Taggable` items can have data attached to them. This provides hash-map-like functionality,
+/// but may be optimized for faster lookup.
+pub trait Taggable<Node, T> {
+    /// Get the index of `node`
+    fn get_tag(&self, node: Node) -> &T;
```

Comment:
Is the reason for not addressing the case where `Node` isn't in the set a matter of efficiency and that your code never tries to get something that doesn't exist?